Port CI from Travis to GitHub actions

MoonlightSentinel opened the pull request.

MoonlightSentinel committed:
Port CI from Travis to GitHub actions

Travis mostly dropped their support for open-Source projects.

MoonlightSentinel committed:
Bump Ubuntu version

MoonlightSentinel committed:
Temporarily disable failing dmd-nightly

MoonlightSentinel committed:
Remove deprecated -disable-fp-elim switch

Geod24 commented on .github/workflows/run-test-sh.yml:
```diff
@@ -0,0 +1,39 @@
+name: run-test-shell
+
+on:
+  pull_request:
+    branches:
+      - master
+      - stable
+
+  push:
+    branches:
+      - master
+      - stable
```
We could just KISS and use `on: [ push, pull_request ]`, no ?

Geod24 commented on .github/workflows/run-test-sh.yml:
```diff
@@ -0,0 +1,39 @@
+
+jobs:
+  main:
+    name: Build and test image
```
I try to keep those names short otherwise you don't see the matrix args in the "Status check" summary.

Geod24 commented on .github/workflows/run-test-sh.yml:
```diff
@@ -0,0 +1,39 @@
+          # - { version: dmd-nightly, exec: dmd   }
+          - { version: dmd-beta,    exec: dmd   }
+          - { version: dmd,         exec: dmd   }
+          - { version: ldc-beta,    exec: ldmd2 }
```
```suggestion
          - { version: ldc-master,    exec: ldmd2 }
```

IMO it's a much better coverage than the infrequent beta. I would advise this for `dmd` too if master wasn't broken.

MoonlightSentinel replied:
Probably but I would rather not deviate that much from the Travis setup during this transition.
Such enhancements should be made as a seperate PR.

PetarKirov replied:
> it's a much better coverage than the infrequent beta

It's actually not about test coverage, but about what options are available in the drop down next to the run button (and obviously in the exec rest api). In general the more options the better... as long as they all work :D

Geod24 commented on .github/workflows/run-test-sh.yml:
```diff
@@ -0,0 +1,39 @@
+      - name: Checkout repo
+        uses: actions/checkout@v2
+        with:
+          submodules: recursive
```
There's only one (flat) submodule, so I feel like `true` should be better ?

Geod24 commented on .github/workflows/run-test-sh.yml:
```diff
@@ -0,0 +1,39 @@
+        with:
+          submodules: recursive
+
+      - name: "Build image"
+        run: docker build -t "dlangtour/core-exec:${{ matrix.version }}" --build-arg "DLANG_VERSION=${{ matrix.version }}" --build-arg "DLANG_EXEC=${{ matrix.exec }}" .
```
Instead of doing it manually, use the docker action.
If you want an example, we do this:
https://github.com/bpfkorea/agora/blob/e5fc8fcd925c81c5fcff354880868b5fdbeffc5b/.github/workflows/agora-docker.yml

MoonlightSentinel commented:
Building and testing works in my fork but tends to spuriously fail due to a broken pipe (idk why).

Upload should work as well once a mantainer adds the appropriate secrets.

MoonlightSentinel commented on Dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM ubuntu:19.10
+FROM ubuntu:latest
```
`19.10` and `20.10` fail due to missing repositorys

PetarKirov replied:
Given that `latest` = LTS, I agree it's a good change.

MoonlightSentinel commented on test.sh:
```diff
@@ -56,11 +56,11 @@ bsource=$(echo $source | base64 -w0)
+#     source='void main() { int a; }'
+#     bsource=$(echo $source | base64 -w0)
+#     DOCKER_FLAGS="-asm" docker run -e DOCKER_FLAGS --rm $dockerId $bsource | grep -q "_Dmain"
+# fi
```
Requires #57

PetarKirov replied:
You can drop this commit, now that #57 is merged.

MoonlightSentinel commented on test.sh:
```diff
@@ -75,13 +75,28 @@ if [[ $dockerId =~ "ldc" ]] ; then
     #                                                        ^ line 4 column 2
     bsource=$(echo "$source" | base64 -w0)
-    DOCKER_FLAGS="-fsanitize=address -g -disable-fp-elim" docker run -e DOCKER_FLAGS --rm $dockerId $bsource | grep -q '#0 0x[0-9a-f]* in ..main [/a-z\.]*:4:2'
+    DOCKER_FLAGS="-fsanitize=address -g" docker run -e DOCKER_FLAGS --rm $dockerId $bsource | grep -q '#0 0x[0-9a-f]* in ..main [/a-z\.]*:4:2'
```
`-disable-fp-elim` was [removed in 1.24.0-beta1](https://github.com/ldc-developers/ldc/issues/3576)

MoonlightSentinel commented on test.sh:
```diff
@@ -75,13 +75,28 @@ if [[ $dockerId =~ "ldc" ]] ; then
+    rm $html_target
+    exit 1
+fi
+rm $html_target
```
This test consistently failed due to a broken pipe on Actions

Geod24 commented:
> Building and testing works in my fork but tends to spuriously fail due to a broken pipe (idk why).

Might be: https://github.com/dlang/dub/pull/1998#issuecomment-694634112 ?

MoonlightSentinel committed:
Replace piped grep with custom function

They apparently trigger errors with broken pipes and a custom function
also allows for a better error reporting

MoonlightSentinel commented:
> Might be: [dlang/dub#1998 (comment)](https://github.com/dlang/dub/pull/1998#issuecomment-694634112) ?

Maybe. Replaced the pipe with a temporary file handled by a dedicated function which doesn't fail spuriously and provides better error messages.

PetarKirov commented:
@MoonlightSentinel Thanks a lot for your work, looks good to me overall! I have merged #57 and rebased your PR on top of master and dropped the commit which disabled the obj2asm test. I have also set the `DOCKER_USERNAME` and `DOCKER_TOKEN` GitHub Actions secrets, so overall, I think we should be good now.

PetarKirov commented:
I think I need to merge this PR, so we can get the first pipeline started on this repo. Let's do that even if there may be issues, as we can easily fix them in a subsequent PR.

PetarKirov commented:
Yep, first pipeline on the repo just started:

![image](https://user-images.githubusercontent.com/3475150/110441824-dcb82b00-80c2-11eb-85e4-e757dbbcd933.png)

PetarKirov commented:
Continuing the work in #59
